File: level2/ex25.js
function solution(orders, courses) {
  // n개의 요소로 구성된 부분집합을 구하는 함수
  function subset(order, n) {
    const answer = [];

    // 부분집합을 구하는 dfs 함수
    function dfs(arr = [], index) {
      if (arr.length === n ) {
        answer.push(arr.join(''));
        return;
      }

      if (index === order.length) {
        return;
      }

      dfs([...arr, order[index]], index + 1);
      dfs([...arr], index + 1);
    }

    dfs([], 0);

    return answer;
  }

  const answer = [];

  // 각 주문을 알파벳을 기준으로 오름차순 정렬
  orders = orders.map(oder => oder.split('').sort().join(''));

  courses.forEach(course => {
    const combinations = {};

    // 각 주문에 대한 부분 메뉴 수 카운팅
    orders.forEach(order => {
      const subsets = subset(order, course);

      subsets.forEach(subset => {
        const keys = Object.keys(combinations);

        if (keys.includes(subset)) combinations[subset]++;
        else combinations[subset] = 1;
      });
    });

    const keys = Object.keys(combinations);
    const values = Object.values(combinations);

    // 조합이 나오지 않는 경우
    if (keys.length === 0) return;

    // 최다 주문 횟수
    const maxOrdered = values.reduce((a, b) => Math.max(a, b));

    // 2번 미만 주문된 조합밖에 없는 경우
    if (maxOrdered <= 1) return;

    // 최다 주문 조합 찾기
    keys.forEach(key => {
      if (combinations[key] === maxOrdered) answer.push(key);
    });
  });

  // 알파벳을 기준으로 오름차순 정렬
  return answer.sort();
}

File: level2/ex25.py
from itertools import combinations

# 각 조합의 주문 횟수를 반환하는 함수 
def get_count(orders, n):
    combs = []
    
    for order in orders:
        combs += list(combinations(order, n))
    
    combs = ["".join(comb) for comb in combs]
    
    return {comb: combs.count(comb) for comb in combs if combs.count(comb) >= 2}

# 최다 주문 조합 반환 함수
def get_max(orders, n):
    count = get_count(orders, n)
    
    if len(count) > 0:
        max_count = max(count.values())
        return [comb for comb in count if count[comb] == max_count]
    else:
        return []
    
def solution(orders, course):
    # 조합
    answer = []

    # 정렬
    orders = [sorted(order) for order in orders]
    
    for n in course:
        answer += get_max(orders, n)
    
    # 정렬
    answer.sort()
    
    return answer